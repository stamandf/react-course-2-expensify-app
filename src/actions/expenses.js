import uuid from 'uuid';

// ADD_EXPENSE Action
export const addExpense = ({description = '', note = '', amount = 0, createdAt = 0 } = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
});

export const removeExpense = ({id} = {}) => ({
    type: 'DELETE_EXPENSE',
    id
});
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});
