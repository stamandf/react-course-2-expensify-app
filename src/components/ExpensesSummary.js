import React from 'react';
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';
import { connect } from 'react-redux';

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
    if (expensesCount === 0) {
        return <p>No expenses.</p>
    } else {
        const expenseWord = expensesCount > 1 ? ("expenses") : ("expense");
        const totalFormatted = numeral(expensesTotal/100).format('$0,0.00'); 
        return (
            <p>
            Viewing {expensesCount} {expenseWord} totalling : {totalFormatted}.
        </p>
        )
        
    }
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expensesCount: visibleExpenses.length, 
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
};
export default connect(mapStateToProps)(ExpensesSummary);

//Previous version:
//================================================

// export const ExpensesSummary = (props) => {
//     console.log("props:", props);
    
//     if (props.expenses.length === 0) {
//         return <p>No expenses.</p>
//     } else {
//         const expenseWord = props.expenses.length > 1 ? ("expenses") : ("expense");
//         const amount = selectExpensesTotal(props.expenses); 
//         return (
//             <p>
//             Viewing {props.expenses.length} {expenseWord} totalling : {numeral(amount/100).format('$0,0.00')}
//         </p>
//         )
        
//     }
// }

// const mapStateToProps = (state) => {
//     return {
//         expenses:  selectExpenses(state.expenses, state.filters), 
//         filters: state.filters
//     };
// };
