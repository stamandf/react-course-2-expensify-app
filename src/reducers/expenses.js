// Expenses reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {  //Action creator
    switch (action.type) {
        case 'ADD_EXPENSE':
            // return state.concat(action.expense);
            return [...state, action.expense]; //The spreading operator for arrays
        case 'DELETE_EXPENSE':
            return state.filter(({id}) => id != action.id);  
        case 'EDIT_EXPENSE':   
            return state.map((expense) => {
                if(expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                };
            });
        default:
            return state;
    }
};
