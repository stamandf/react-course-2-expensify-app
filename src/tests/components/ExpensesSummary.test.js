import React from 'react';
import { shallow } from 'enzyme';
import ExpensesSummary from '../../components/ExpensesSummary';
import selectExpensesTotal from '../../selectors/expenses-total';
import expenses from '../fixtures/expenses';

test('should display multiple expenses summary correctly', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();

    // const output = selectExpensesTotal(expenses);
    // expect(output).toBe("Viewing 3 expenses totalling : $1,141.95"); //1,141.95
})

test('should display one expense summary correctly', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses[2]} />);
    expect(wrapper).toMatchSnapshot();

});

test('should display expense summary correctly when no expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[]} />);
    expect(wrapper).toMatchSnapshot();

});
