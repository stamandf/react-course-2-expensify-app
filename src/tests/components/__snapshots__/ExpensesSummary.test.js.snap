// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display expense summary correctly when no expenses 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`should display multiple expenses summary correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`should display one expense summary correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
